### PR TITLE
Jetpack plugin: Remotely check for partner coupon validity

### PR DESCRIPTION
**Pull request by kallehauge**
#### Changes proposed in this Pull Request:

We wish to be able to more "aggressively" push partner coupons, if they exist in the WordPress site's DB.
A prerequisite for this is having more up-to-date information about a partner coupon than "just purging it after 30 days" because the 30 day delay would mean that we keep highlighting coupons for 30 days even if they are: used, invalid, from a deactivated partner. 

#### Jetpack product discussion

* pdpAdu-1A-p2
* 1201265858818217-as-1201673004210721

#### Does this pull request change what data or activity we track or use?

No.

#### Testing instructions:

* Testing instructions can be found at D73431-code

### CODE REVIEW

File: projects/packages/partner/src/class-partner-coupon.php [hunk 1]
```diff
@@ -127,27 +137,125 @@ public function catch_coupon( $plugin_slug, $redirect_location ) {
...
+	 */
+	protected function maybe_purge_coupon_by_availability_check( $connection_client = null ) {
+		// Allow for client dependency injection to allow for mock client during unit testing.
+		if (
+			empty( $connection_client ) ||
+			! class_exists( $connection_client ) ||
+			! method_exists( $connection_client, 'wpcom_json_api_request_as_blog' )
+		) {
+			$connection_client = Connection_Client::class;
+		}
```
Comment (atanas-dev) [inline projects/packages/partner/src/class-partner-coupon.php:212]: I think it would be better if we inject a callable in the class constructor instead e.g.:
```php
<?php
/**
 * Callable that executes a blog-authenticated request.
 *
 * @var callable
 */
protected $request_as_blog;

/**
 * Get singleton instance of class.
 *
 * @return Partner_Coupon
 */
public static function get_instance() {
    if ( is_null( self::$instance ) ) {
        self::$instance = new Partner_Coupon( [ Connection_Client::class, 'wpcom_json_api_request_as_blog' ] );
    }

    return self::$instance;
}

/**
 * Constructor.
 *
 * @param callable $request_as_blog Callable that executes a blog-authenticated request.
 */
public function __construct( $request_as_blog ) {
    $this->request_as_blog = $request_as_blog;
}
```

And then in this method we'd run this:
```php
$response = $this->$request_as_blog(
    sprintf(
        '/sites/%d/jetpack-partner/coupon/v1/site/coupon?coupon_code=%s',
        $jetpack_id,
        $coupon['coupon_code']
    ),
    2,
    array( 'method' => 'GET' ),
    null,
    'wpcom'
);
```

This way:
- we're not coupled to `Connection_Client::wpcom_json_api_request_as_blog()`
- the method that makes the request doesn't have to do parameter validation
- if we ever need to make a request in another method we don't have to pass the parameter again
- tests would be simpler (e.g. we won't need the Connection_Client mock class.)

File: projects/packages/partner/src/class-partner-coupon.php [hunk 2]
```diff
@@ -127,27 +137,125 @@ public function catch_coupon( $plugin_slug, $redirect_location ) {
...
+		if ( ! $jetpack_id ) {
+			return false;
+		}
+
+		$coupon   = self::get_coupon();
+		$response = $connection_client::wpcom_json_api_request_as_blog(
+			sprintf(
+				'/sites/%d/jetpack-partner/coupon/v1/site/coupon?coupon_code=%s',
+				$jetpack_id,
+				$coupon['coupon_code']
```
Comment (atanas-dev) [inline projects/packages/partner/src/class-partner-coupon.php:224]: Let's use add_query_arg() rather than stitch query params.

File: projects/packages/partner/src/class-partner-coupon.php [hunk 3]
```diff
@@ -127,27 +137,125 @@ public function catch_coupon( $plugin_slug, $redirect_location ) {
...
+		// Allow for client dependency injection to allow for mock client during unit testing.
+		if (
+			empty( $connection_client ) ||
+			! class_exists( $connection_client ) ||
+			! method_exists( $connection_client, 'wpcom_json_api_request_as_blog' )
+		) {
+			$connection_client = Connection_Client::class;
+		}
+
+		$jetpack_id = Jetpack_Options::get_option( 'id', false );
```
Comment (atanas-dev) [inline projects/packages/partner/src/class-partner-coupon.php:214]: I think we should stick to `$blog_id`.

File: projects/packages/partner/src/class-partner-coupon.php [hunk 4]
```diff
@@ -127,27 +137,125 @@ public function catch_coupon( $plugin_slug, $redirect_location ) {
...
+			array( 'method' => 'GET' ),
+			null,
+			'wpcom'
+		);
+
+		$body = json_decode( wp_remote_retrieve_body( $response ), true );
+
+		if (
+			200 === wp_remote_retrieve_response_code( $response ) &&
+			isset( $body['available'] ) &&
```
Comment (atanas-dev) [inline projects/packages/partner/src/class-partner-coupon.php:236]: This seems like it can raise an error if the decoded value is not an array - isset might avoid it, but it won't hurt to add an is_array() check as well. Leaving this entirely up to you.

File: projects/packages/partner/src/class-partner-coupon.php [hunk 5]
```diff
@@ -127,27 +137,125 @@ public function catch_coupon( $plugin_slug, $redirect_location ) {
...
+	 */
+	protected function delete_coupon_data() {
+		Jetpack_Options::delete_option(
+			array(
+				self::$coupon_option,
+				self::$added_option,
+			)
+		);
+
+		Jetpack_Options::delete_raw_option( self::$last_check_option );
```
Comment (atanas-dev) [inline projects/packages/partner/src/class-partner-coupon.php:258]: Could you please let me know why this option needs to be dealt with by using the raw methods? I'm not sure I understand what's at work here.

Comment (kallehauge) [reply projects/packages/partner/src/class-partner-coupon.php:258]: "Raw" means that it's actually stored as its own option row in `wp_options` vs non-raw which are stores in the same `jetpack_options` row in the same table - but moving it to a transient makes this mute in our case

File: projects/packages/partner/composer.json [hunk 6]
```diff
@@ -3,12 +3,14 @@
...
 	"license": "GPL-2.0-or-later",
-	"require": {},
+	"require": {
+		"automattic/jetpack-connection": "^1.36",
+		"automattic/jetpack-options": "^1.14"
+	},
 	"require-dev": {
 		"brain/monkey": "2.6.1",
-		"yoast/phpunit-polyfills": "1.0.3",
+		"yoast/phpunit-polyfills": "1.0.2",
```
Comment (kallehauge) [inline projects/packages/partner/composer.json:12]: `"automattic/jetpack-connection": "^1.36"` required us to downgrade `yoast/phpunit-polyfills` to `1.0.2`

File: projects/packages/partner/src/class-partner-coupon.php [hunk 7]
```diff
@@ -127,27 +137,125 @@ public function catch_coupon( $plugin_slug, $redirect_location ) {
...
+	 * errors.
 	 */
-	public function purge_coupon() {
+	public function maybe_purge_coupon() {
+		// Only run coupon checks on Jetpack admin pages.
+		// The "admin-ui" package is responsible for registering the Jetpack admin
+		// page for all Jetpack plugins and has hardcoded the settings page to be
+		// "jetpack", so we shouldn't need to allow for dynamic/custom values.
+		// phpcs:ignore WordPress.Security.NonceVerification.Recommended
+		if ( ! isset( $_GET['page'] ) || ! 'jetpack' === $_GET['page'] ) {
```
[the remark this answers is not in the document]
Comment (kallehauge) [reply projects/packages/partner/src/class-partner-coupon.php]: Ha! Doh! Thanks 😂

File: projects/packages/partner/src/class-partner-coupon.php [hunk 8]
```diff
@@ -127,27 +137,125 @@ public function catch_coupon( $plugin_slug, $redirect_location ) {
...
+		}
+
+		if ( $this->maybe_purge_coupon_by_added_date() ) {
+			return;
+		}
+
+		$last_check = Jetpack_Options::get_raw_option( self::$last_check_option, 0 );
+
+		// Limit checks to happen every minute instead of on all page requests.
+		if ( $last_check && time() > $last_check + MINUTE_IN_SECONDS ) {
```
[the remark this answers is not in the document]
Comment (kallehauge) [reply projects/packages/partner/src/class-partner-coupon.php]: Agreed - I actually forgot they were a thing 😅  I've converted it to a transient (which also removed the need to update the options package 👏  ).

File: projects/packages/partner/src/class-partner-coupon.php (hunk 1 above)
Comment (kallehauge) [reply projects/packages/partner/src/class-partner-coupon.php]: It also just fits better with DI, so a great suggestion! 👍

Comment (atanas-dev) [reply projects/packages/partner/src/class-partner-coupon.php]: That's where the inspiration came from - yes 😁

File: projects/packages/partner/src/class-partner-coupon.php [hunk 9]
```diff
@@ -127,27 +153,123 @@ public function catch_coupon( $plugin_slug, $redirect_location ) {
...
+		// phpcs:ignore WordPress.Security.NonceVerification.Recommended
+		if ( ! isset( $_GET['page'] ) || 'jetpack' !== $_GET['page'] ) {
+			return;
+		}
+
+		if ( $this->maybe_purge_coupon_by_added_date() ) {
+			return;
+		}
+
+		// Limit checks to happen every minute instead of on all page requests.
```
Comment (atanas-dev) [inline projects/packages/partner/src/class-partner-coupon.php:175]: I'd reword this to "Limit checks to happen once a minute at most." otherwise this sounds like it executes every minute like a cron job.

File: projects/packages/partner/src/class-partner-coupon.php [hunk 10]
```diff
@@ -38,6 +39,20 @@ class Partner_Coupon {
 	 */
 	public static $added_option = 'partner_coupon_added';
 
+	/**
+	 * Name of "last availability check" transient.
+	 *
+	 * @var string
+	 */
+	public static $last_check_option = 'jetpack_partner_coupon_last_check';
```
Comment (atanas-dev) [inline projects/packages/partner/src/class-partner-coupon.php:47]: How about `$last_check_transient`?

File: projects/packages/partner/tests/php/test-partner-coupon.php [hunk 11]
```diff
@@ -229,4 +232,94 @@ public function dataprovider_purge_dates() {
...
+		}
+
+		$mock_client = $this->getMockBuilder( \stdClass::class )
+							->setMethods( array( 'wpcom_json_api_request_as_blog' ) )
+							->getMock();
+
+		$mock_client
+			->expects( $this->once() )
+			->method( 'wpcom_json_api_request_as_blog' )
+			->willReturn( $mock_response );
```
Comment (atanas-dev) [inline projects/packages/partner/tests/php/test-partner-coupon.php:266]: We can use an anonymous function instead of dealing with the mock builder:
```
$mock_requester = function () use ( $mock_response ) { return $mock_response; };
$instance = new Partner_Coupon( $mock_requester );
```

**Review by atanas-dev — DISMISSED** [review]
Tests good for me :+1:
I've just posted some super minor non-blocking stuff.

---

We might want to run `maybe_purge_coupon_by_availability_check()` before doing the showCouponRedemption redirect so we don't redirect to it if the coupon is not actually available.
The current redirect doesn't really break anything and using `maybe_purge_coupon_by_availability_check()` will introduce a remote request delay but I'm slightly leaning towards not doing the redirect in cases where the coupon is invalid.

File: projects/packages/partner/tests/php/test-partner-coupon.php (hunk 11 above)
Comment (kallehauge) [reply projects/packages/partner/tests/php/test-partner-coupon.php:266]: It's is surely much simpler to read!
I would prefer to keep the `expects( $this->once() )` check since we make a remote request that will block the page from loading until it's done. If we - for any reason - make the request more than once, then it's 1) an error and 2) a bad user experience.

Although the above could be argued to belong in its own test, but lets keep it simple 😄

**Comment by kallehauge** [comment]
> We might want to run `maybe_purge_coupon_by_availability_check()` before doing the showCouponRedemption redirect so we don't redirect to it if the coupon is not actually available.
> The current redirect doesn't really break anything and using `maybe_purge_coupon_by_availability_check()` will introduce a remote request delay but I'm slightly leaning towards not doing the redirect in cases where the coupon is invalid.

If we don't redirect the user, then they would most likely end up on the WordPress Admin dashboard instead which, IMO, doesn't change the experience for the user. So the fact that we redirect them to the Jetpack dashboard no matter what seems OK to me.
But you do make a very valid point that's worth discussing in a bit bigger context: Since we're now able to check the availability of a coupon, creating a new component/page to inform the user "Hey, the coupon you just tried to claim isn't available" instead of just nothing happening is probably something we should have in the next iteration, because no matter if/where we redirect, there's a dislink is what the user expected vs nothing happening after they've clicked a "Claim" link from a partner.

Let's consider it outside of this tasks scope for now (I've created a task so we can talk to design about the new flow opportunity 1201265858818217-as-1201802907668685) 😄

**Comment by kallehauge** [comment]
For crew: I broke Atanas' PR approval after addressing a variable name + fixed a composer conflict, so consider it ready for crew review 😄

File: projects/packages/partner/src/class-partner-coupon.php [hunk 12]
```diff
@@ -127,27 +153,123 @@ public function catch_coupon( $plugin_slug, $redirect_location ) {
...
+	public function maybe_purge_coupon() {
+		// Only run coupon checks on Jetpack admin pages.
+		// The "admin-ui" package is responsible for registering the Jetpack admin
+		// page for all Jetpack plugins and has hardcoded the settings page to be
+		// "jetpack", so we shouldn't need to allow for dynamic/custom values.
+		// phpcs:ignore WordPress.Security.NonceVerification.Recommended
+		if ( ! isset( $_GET['page'] ) || 'jetpack' !== $_GET['page'] ) {
+			return;
+		}
+
```
Comment (jeherve) [inline projects/packages/partner/src/class-partner-coupon.php:170]: I wonder if we should shortcircuit early when the site is not connected or in offline mode, to avoid doing too much work when it's not necessary?

**Review by jeherve — DISMISSED** [review]
This tests well for me. It should be good to merge. I have a small remark below, but I think that could be addressed in a follow-up PR if you'd like. I'll leave that up to you.

**Review by jeherve — COMMENTED** [review]
I've given it another try with the changes, thank you! I don't know if I'm doing something wrong, but for some reason I cannot get the purge to happen anymore; the expired coupon stays on the site all the time. Does it work for you?

Also linking #22413 here for my own cross-reference, since it will fix some of the issues with offline mode and the coupon redeeming UI showing up.

**Comment by kallehauge** [comment]
> I've given it another try with the changes, thank you! I don't know if I'm doing something wrong, but for some reason I cannot get the purge to happen anymore; the expired coupon stays on the site all the time. Does it work for you?

I just tested and it works for me @jeherve. You're probably doing it correctly and it's something else, but:

1. If you tested with the working coupon first, did you then wait 60s before testing the expired coupon?
2. Have you defined `define( 'JETPACK_DEV_DEBUG', true );` in `0-sandbox.php`?
3. Did you sandbox the API? The coupon will only be removed if we have a "successful" response that says it should be removed but it will not inform you about any errors during the look-up.

If all of the above is OK, would you mind trying to replace the new conditions and see if you get a blank page with one of the messages, for any reason?

```
if ( ( new Status() )->is_offline_mode() ) {
	var_dump('Its in offline mode');
	die;
	return;
}

$connection = new Connection_Manager( $plugin_slug );
if ( ! $connection->is_connected() ) {
	var_dump('The site is not connected');
	die;
	return;
}
```

If this still doesn't give you anything, then I would love to see your body and status code in `maybe_purge_coupon_by_availability_check` _(Feel free to disable the transient while testing this, so you don't have to wait 60s between tries)_.

It's quite a few hoops you might have to jump through, so feel free to DM me so we can rubber duck what might be going on 😄

**Review by samiff — APPROVED** [review]
I was able to see valid coupon option added, and the expired one being removed okay. Not sure if @jeherve was running into something else, but it should be easier to test now that the wpcom diff has been merged.

**Comment by kallehauge** [comment]
> I was able to see valid coupon option added, and the expired one being removed okay.

Sounds great! Thanks for testing @samiff 😄  

> Not sure if jeherve was running into something else, but it should be easier to test now that the wpcom diff has been merged.

He did succeed in his initial test, so I'm pretty curious as well. Jurassic Ninja has been teasing with this branch in beta mode, so it could have been that he started a JN which used `stable` instead of the branch 🤷
I've tested it pretty rigorously after his report and it seems OK, so I'm 🤞  😄
